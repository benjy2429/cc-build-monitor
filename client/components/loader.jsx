import React from 'react';
import fetchData from '../providers/circleci.js';

export default Component => (
    class extends React.Component {
        constructor(props) {
            super(props);
            this.state = {
                loading: true,
                data: {}
            };
        }

        componentDidMount() {
            fetchData().then(data => {
                this.setState({ loading: false, data });
            });
        }

        render() {
            const { loading, data } = this.state;

            return loading ?
                <div>Loading...</div> :
                <Component {...this.props} data={data} />;
        }
    }
);
