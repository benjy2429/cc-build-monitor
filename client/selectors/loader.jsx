import React from 'react';
import axios from 'axios';
import Overlay from '../components/overlay';

const refreshRate = 60000;
const defaultFetch = () => axios({
  url: '/fetch',
  timeout: 5000,
});

export default (Component, fetch = defaultFetch) => (
  class Loader extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        loading: true,
        error: false,
        data: null,
      };
    }

    async fetchData() {
      this.setState({ loading: true });
      try {
        const res = await fetch();
        if (res.data.error) {
          this.setState({ loading: false, error: res.data.error });
          return;
        }
        this.setState({ loading: false, error: false, data: res.data });
      } catch (e) {
        this.setState({ error: e.message, loading: false });
      }
    }

    componentDidMount() {
      this.fetchData();
      this.intervalId = setInterval(this.fetchData.bind(this), refreshRate);
    }

    componentWillUnmount() {
      clearInterval(this.intervalId);
    }

    render() {
      const { loading, error, data } = this.state;

      return (
        <div className="full-height">
          { error && <Overlay title="Error fetching data" subtitle={error} /> }
          { loading && <div className="loading-spinner" /> }
          { data && <Component {...this.props} {...data} /> }
        </div>
      );
    }
  }
);
