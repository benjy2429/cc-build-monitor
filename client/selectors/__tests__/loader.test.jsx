import React from 'react';
import { mount } from 'enzyme';
import loader from '../loader';

const WrappedComponent = () => <div className="wrapped-component" />;
const defaultStubFetch = () => Promise.resolve({ data: { some: 'data' } });

const renderLoader = (stubFetch = defaultStubFetch) => {
  const LoaderComponent = loader(WrappedComponent, stubFetch);
  return mount(<LoaderComponent />);
};

describe('Loader', () => {
  it('renders the spinner initially', () => {
    const component = renderLoader();
    expect(component).toMatchSnapshot();
  });

  it('renders the component after receiving data', async () => {
    const component = renderLoader();
    await defaultStubFetch;
    expect(component).toMatchSnapshot();
  });

  it('renders the component and spinner when re-fetching', async () => {
    jest.useFakeTimers();
    const component = renderLoader();
    await defaultStubFetch;
    jest.runOnlyPendingTimers();
    expect(component).toMatchSnapshot();
  });

  it('renders an overlay for received errors', async () => {
    const stubFetch = () => Promise.resolve({ data: { error: 'received error' } });
    const component = renderLoader(stubFetch);
    await stubFetch;
    expect(component).toMatchSnapshot();
  });

  it('renders an overlay for fetch errors', async () => {
    const stubFetch = () => { throw new Error('fetch error'); };
    const component = renderLoader(stubFetch);
    await stubFetch;
    expect(component).toMatchSnapshot();
  });
});
