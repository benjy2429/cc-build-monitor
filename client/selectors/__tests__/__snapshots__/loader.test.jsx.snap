// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader renders an overlay for fetch errors 1`] = `
<Loader>
  <div
    className="full-height"
  >
    <Overlay
      subtitle="fetch error"
      title="Error fetching data"
    >
      <div
        className="overlay"
      >
        <div
          className="overlay-wrapper"
        >
          <div
            className="overlay-title"
          >
            Error fetching data
          </div>
          <div
            className="overlay-subtitle"
          >
            fetch error
          </div>
        </div>
      </div>
    </Overlay>
  </div>
</Loader>
`;

exports[`Loader renders an overlay for received errors 1`] = `
<Loader>
  <div
    className="full-height"
  >
    <Overlay
      subtitle="received error"
      title="Error fetching data"
    >
      <div
        className="overlay"
      >
        <div
          className="overlay-wrapper"
        >
          <div
            className="overlay-title"
          >
            Error fetching data
          </div>
          <div
            className="overlay-subtitle"
          >
            received error
          </div>
        </div>
      </div>
    </Overlay>
  </div>
</Loader>
`;

exports[`Loader renders the component after receiving data 1`] = `
<Loader>
  <div
    className="full-height"
  >
    <WrappedComponent
      some="data"
    >
      <div
        className="wrapped-component"
      />
    </WrappedComponent>
  </div>
</Loader>
`;

exports[`Loader renders the component and spinner when re-fetching 1`] = `
<Loader>
  <div
    className="full-height"
  >
    <div
      className="loading-spinner"
    />
    <WrappedComponent
      some="data"
    >
      <div
        className="wrapped-component"
      />
    </WrappedComponent>
  </div>
</Loader>
`;

exports[`Loader renders the spinner initially 1`] = `
<Loader>
  <div
    className="full-height"
  >
    <div
      className="loading-spinner"
    />
  </div>
</Loader>
`;
